### PR TITLE
ctest targets from catch2 and fastcov

### PR DESCRIPTION
**Pull request by dlyr**
* **Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines
- [ ] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
Modify cmake conf to create ctest from catch2 tests.
Also add a cmake module to manage coverage target, in fact our old behavior is fine/better for lcov, but fastcov is really cool
I prefer to keep lcov for codecov.io reports (since it works) but local coverage is _way_ faster with fastcov.

* **What is the current behavior?** (You can also link to an open issue here)
for tests, a ad hoc call to unitest is added

* **What is the new behavior (if this is a feature change)?**
now make test or ctest show each catch2 test exec, and can be done in parallel

### CODE REVIEW

**Review by nmellado — DISMISSED** [review]
Ready to merge once updated

**Review by nmellado — COMMENTED** [review]
I've been thinking, could we add in the documentation how to compute the coverage locally ?

**Comment by dlyr** [comment]
Sure, but it's only tested on my computer ;) don't know how it scales to macos and windows for instance.

**Comment by nmellado** [comment]
Add least add the procedure and I'll see if I can replicate it on macOS.

**Comment by dlyr** [comment]
done.

File: doc/developer/develWithRadium.md [hunk 1]
```diff
@@ -38,3 +39,39 @@ If your test needs data, add them to `tests/unittest/data`. unittest working dir
 Here is an example:
 
 \snippet unittest/Core/obb.cpp obb test
+
+# Code coverage
+
+On PR, code coverage analysis is perform through a github action with help of codecov.io report.
```
Comment (MathiasPaulin) [inline doc/developer/develWithRadium.md:45]:
```suggestion
On PR, code coverage analysis is performed through a github action with help of codecov.io report.
```

File: doc/developer/develWithRadium.md [hunk 2]
```diff
@@ -38,3 +39,39 @@ If your test needs data, add them to `tests/unittest/data`. unittest working dir
...
+
+Then build Radium with coverage enable, run test and generate report.
+
+```bash
+#some vars
+RADIUMC_CONFIG=path/to/file/with/radium-options.cmake #lik where external are built ... it's optionnal
+RADIUM_SOURCE=path/to/radium/main/CMakeLists.txt
+
+# in you build path
+cmake -DCMAKE_BUILD_TYPE=Debug -DRADIUM_ENABLE_COVERAGE=ON -C RADIUM_CONFIG RADIUM_SOURCE
```
[the remark this answers is not in the document]
Comment (dlyr) [reply doc/developer/develWithRadium.md:58]:
```suggestion
cmake -DCMAKE_BUILD_TYPE=Debug -DRADIUM_ENABLE_COVERAGE=ON -C ${RADIUM_CONFIG} ${RADIUM_SOURCE}
```

File: doc/developer/develWithRadium.md [hunk 3]
```diff
@@ -38,3 +39,39 @@ If your test needs data, add them to `tests/unittest/data`. unittest working dir
...
+On PR, code coverage analysis is perform through a github action with help of codecov.io report.
+You can run such analysis on your own (on linux system, not tested on other plateform) using lcov.
+First install `lcov`, and ` libxml2-utils` for the tests.
+It's propably on your system repositories, for Debian for instance `sudo apt install lcov libxml2-utils`.
+
+Then build Radium with coverage enable, run test and generate report.
+
+```bash
+#some vars
+RADIUMC_CONFIG=path/to/file/with/radium-options.cmake #lik where external are built ... it's optionnal
```
[the remark this answers is not in the document]
Comment (dlyr) [reply doc/developer/develWithRadium.md:54]:
```suggestion
RADIUM_CONFIG=path/to/file/with/radium-options.cmake #like where external are built ... it's optionnal
```

File: doc/developer/develWithRadium.md [hunk 4]
```diff
@@ -38,3 +39,39 @@ If your test needs data, add them to `tests/unittest/data`. unittest working dir
 Here is an example:
 
 \snippet unittest/Core/obb.cpp obb test
+
+# Code coverage
+
+On PR, code coverage analysis is perform through a github action with help of codecov.io report.
+You can run such analysis on your own (on linux system, not tested on other plateform) using lcov.
+First install `lcov`, and ` libxml2-utils` for the tests.
```
[the remark this answers is not in the document]
Comment (dlyr) [reply doc/developer/develWithRadium.md]: replaced by lcov, xmllint and sed.

File: doc/developer/develWithRadium.md [hunk 5]
```diff
@@ -38,3 +39,39 @@ If your test needs data, add them to `tests/unittest/data`. unittest working dir
...
+cmake -DCMAKE_BUILD_TYPE=Debug -DRADIUM_ENABLE_COVERAGE=ON -C ${RADIUM_CONFIG} ${RADIUM_SOURCE}
+make lcov #build everything, with coverage enable, run test, perform analysis
+make lcov-report # generate html report
+```
+output report is `./lcov/index.html`
+
+`lcov` is quite slow to process files, [`fastcov`](https://github.com/RPGillespie6/fastcov/blob/master/fastcov.py) is a parallel alternative.
+To install fastcov `pip3 install fastcov`
+
+and for the analysiss
```
Comment (nmellado) [inline doc/developer/develWithRadium.md:67]:
```suggestion
and for the analysis
```

**Review by nmellado — APPROVED** [review]

**Comment by nmellado** [comment]
@dlyr coverage is already introduced in `buildchain.md`: https://github.com/dlyr/Radium-Engine/blob/coverage/doc/developer/buildchain.md

I don't know which place is best, I don't mind, but we should avoid to duplicate.

**Comment by dlyr** [comment]
Right @nmellado , let's say "How to contribute to Radium Libraries" is the best place, we expect PR to have tests and coverage ...
I have merged the two docs (and I didn't remember there were already some doc about coverage ...

**Comment by nmellado** [comment]
I would suggest make cmake fail when the user request coverage but without the right compiler.
Otherwise the message is lost in the cmake log, and it is not easy to understand why `make` does not have a `lcov` target.
People should only activate this option in purpose.
What do you think @dlyr  ?

```diff
-         message( WARNING "[Tests] Coverage requested (RADIUM_ENABLE_COVERAGE), but not activated.")
-         message( WARNING "[Tests] Coverage only supported in DEBUG with GCC.")
+         message( ERROR "[Tests] Coverage requested (RADIUM_ENABLE_COVERAGE), but not activated.")
+         message( ERROR "[Tests] Coverage only supported in DEBUG with GCC.")
```
 https://github.com/STORM-IRIT/Radium-Engine/pull/761/files#diff-8c08cffd7bc74857c105f1ddeeeac9dec123aa93878fce4bff1686925eee2956L88-L89

**Review by nmellado — APPROVED** [review]

**Pull request merged**